extract_data/metadata.py

```diff
@@ -866,3 +1121,204 @@ def change_model_title(model_id, new_title):
+    for x in range(1, 1 + len(s1)):
+        for y in range(1, 1 + len(s2)):
+            if s1[x - 1] == s2[y - 1]:
+                m[x][y] = m[x - 1][y - 1] + 1
+                if m[x][y] > longest:
+                    longest = m[x][y]
+                    x_longest = x
+            else:
+                m[x][y] = 0
+    return s1[x_longest - longest : x_longest]
+
+
+import requests
+
+
+def fetch_crossref_data(doi):
+    url = f"https://api.crossref.org/works/{doi}/transform/application/vnd.crossref.unixsd+xml"
+    try:
+        response = requests.get(url)
+        response.raise_for_status()  # Raise an error if the HTTP request failed
+        if response.status_code == 200:
+            return response.text  # Return the XML content as a string
+        else:
+            print(
+                f"Error fetching data from Crossref for DOI {doi}: Status code {response.status_code}"
+            )
+            return None
+    except requests.exceptions.RequestException as e:
+        print(f"Request error for DOI {doi}: {e}")
+        return None
+
+
+from bs4 import BeautifulSoup
+
+
+def parse_crossref_xml(xml_data):
+    soup = BeautifulSoup(xml_data, "xml")
+    title = soup.find("title").text if soup.find("title") else ""
+    title = title.lower().capitalize()
+
+    authors = soup.find("surname").text if soup.find("surname") else ""
+    authors = authors.lower().capitalize()
+
+    journal = soup.find("full_title").text if soup.find("full_title") else ""
+    publication_date = soup.find("publication_date", media_type="print")
+    if publication_date:
+        year = (
+            publication_date.find("year").text if publication_date.find("year") else ""
+        )
+    else:
+        publication_date = soup.find("publication_date", media_type="online")
+        if publication_date:
+            year = (
+                publication_date.find("year").text
+                if publication_date.find("year")
+                else ""
+            )
+        else:
+            year = None
+
+    crossref_data = {
+        "title": title,
+        "authors": authors,
+        "journal": journal,
+        "year": year,
+    }
+    return crossref_data
+
+
+def search_papers_by_year(year):
+    return sdb.papers.find({"year.value": year})
+
+
+import tqdm
+import time
+
+
+def find_matching_paper_doi(doi):
+    max_common_substring = ""
+    max_paper = None
+    xml = fetch_crossref_data(doi)
+    crossref_data = parse_crossref_xml(xml)
+    year = crossref_data.get("year", "")
+    papers = search_papers_by_year(year)
+
+    for paper in tqdm.tqdm(papers):
+        paper_text = " ".join(
+            str(paper.get(field, {}).get("value", [{}])[0].get("object_name", ""))
+            if field == "authors" and paper.get(field, {}).get("value", [{}])
+            else str(paper.get(field, ""))
+            for field in ["title", "journal", "authors", "year"]
+        )
+
+        crossref_text = " ".join(
```

Comment:
trivial comment: `crossref_text` is the same thing every pass of the loop, so it should be defined before the loop